Add existing guidelines to qiskit docs guidelines (#113)
Qiskit/qiskit_sphinx_theme: docs_guidelines/apidocs/tutorials.guidelines.rst

####################
Tutorials Guidelines
####################

1. Get the user started
=======================

> 1.1 Lower the barrier of entry 

Reviewer: ```suggestion
1.1 Lower the barrier to entry 
```

> 1.1 Lower the barrier of entry 

Reviewer: I think the headings hierarchy looks odd style-wise when it's rendered in the browser. I think maybe the 1. should have `----` underneath and the 1.1, 1.2 etc. should just be bold

------------------------------

Lower the barrier of entry as much as possible. Assume the user does not know much about what is
covered in the tutorial. If we are expecting users to know something, make it explicit and link them
to the needed explanations (See :ref:`tutorial-guidelines-minimum-explanation`) Avoid jargon as much as possible. Provide a minimum
necessary explanation when the jargon is introduced. (See :ref:`tutorial-guidelines-minimum-explanation`)
 
1.2 Provide setup / prerequisites

> ---------------------------------

Reviewer: Should there be text under headings 1.2 and 1.3? I think a bit more explanation about what you mean by setup/pre-reqs would be good

> 1.3 Add sufficient comments to code

Reviewer: Maybe explain why comments are important?

-----------------------------------

2. Provide a complete picture before they start
===============================================

> 2.1 Provide overview at the start

Reviewer: ```suggestion
2.1 Provide an overview at the start
```

---------------------------------

> Summarize what the tutorial is about in the first paragraph, ideally the first sentence

Reviewer: ```suggestion
Summarize what the tutorial is about in the first paragraph, ideally the first sentence.
```

    "In this tutorial, you will..."

2.2 Provide context
-------------------

> Add a context for why you want to do this tutorial? Or who is this tutorial for?

Reviewer: ```suggestion
Add context to help the user understand the purpose of this tutorial. Why would you want to do this tutorial? Or who is this tutorial for?
```

2.3 Summarize at the end
------------------------

Summarize what the user has accomplished at the end of the tutorial

    "You have built a secure, three-layer hylomorphic stasis engine…" Describe (and admire, in a 

>     mild way) what your learner has accomplished (note - not: “you have learned…”)

Reviewer: I think in this sentence we could elaborate a bit more about not using condescending language. So explain that they shouldn't make assumptions about what the audience has achieved (i.e. explain why not to use "you have learned..") or make assumptions about what they already know or are able to understand (so avoid words like, "obviously", "trivial", or "as you know")

Reviewer 2: When I initially read that point (taken verbatim from [Diataxis](https://diataxis.fr/tutorials/#the-language-of-tutorials)) what I understood was mostly to think of the user as an independent being doing things on their own (in a "you can do it" sense instead of "you are alone with the lions, fight for your life") instead of a pupil that is listening to our lesson ("you have learnt ..."). It's the same reason we use "you" instead of "we", that is, to highlight the agency of the user.

On the other side, you can't admire too much that, for example, a user has built a basic circuit with Qiskit because it would sound as if we were almost mocking them.

However I agree that we'd better explicitly tell writers to avoid any kind of condescending language like "obviously", "trivial" or "as you (should) know". This kind of language can only discourage people from using Qiskit.

3. Ensure that the tutorial works reliably and repeatable
=========================================================

3.1 Test tutorials regularly
----------------------------

> on supported platform and python / Qiskit versions (for example, via CI

Reviewer: this reads like a note, i think it should be a full sentence. Also explain what CI is otherwise we aren't practicing what we preach about jargon haha

Reviewer 2: These guidelines are supposed to be mainly for the mantainers of a Qiskit package. I think it's fair to assume they know what CI is. Also don't forget that this is not supposed to be a tutorial (so jargon is not so big a problem) but I can add an explanation about CI. What do you think? @javabster @HuangJunye

Reviewer: yeah most of the time i think it's fine to assume that, it's just that i was reading this straight after the section when you talk about not using jargon etc. and it just felt a bit hypocritical 😅

Reviewer: i wouldn't block merging on this bc of the acronym, just a suggestion. I do think it should be a full sentence though, and generally i think we should keep the grammar consistent in the page. so either all short note form or all full sentences, not a mix of both

tests)

4. Ensure the user sees results immediately
===========================================

4.1 Every step should produce a result
--------------------------------------

Every step the learner follows should produce a comprehensible result, however small. It can 
be a print statement showing a variable, a plot, etc.
        
    "The output should look something like this…"

4.2 Each step should be brief
-----------------------------
Ideally each step should not take longer than a minute. If it is longer than that, please consider

> ways to show intermediate results of the step, otherwise users do not start worrying about
> somethings are going wrong.

Reviewer: ```suggestion
ways to show intermediate results of the step, or at least make them aware that it takes a while, otherwise users will start worrying that
something is going wrong.
```

5. Don't try to teach
=====================

5.1 Provide concrete steps
--------------------------

> Describe concrete steps, not abstract concepts, ignore options and alternatives

Reviewer: ```suggestion
Describe concrete steps, not abstract concepts, ignore options and alternatives.
```

5.2 Use titles or transition sentences to make steps more obvious
-----------------------------------------------------------------
    
Titles for steps should be an action not a noun.

    "First, do x. Now, do y. Now that you have done y, do z."

5.3 Do not generalize
---------------------

Focus on the particular instance of the case you want to demonstrate, do not generalize / provide
abstractions.

.. _tutorial-guidelines-minimum-explanation:

6. Offer only minimum, necessary, explanation
=============================================

6.1 Explain exactly what's needed
---------------------------------

Try to explain exactly what's needed to get the steps done AND for the user understand

> everything they're doing and why they are doing it in the first place. No more no less.  

Reviewer: ```suggestion
everything they're doing and why they are doing it in the first place. No more, no less.  
```

6.2 Provide links if needed
---------------------------

Link to further explanatory material if and only if it's really needed. Try not to add too

> many links. Because they will interrupt the flow of a tutorial by digressing into explanation.

Reviewer 2: ```suggestion
many links because they will interrupt the flow of a tutorial by digressing into explanation.
```

Reviewer: Do we really want to encourage less links? I feel like we want to encourage our devs to use links more, and this para might have the opposite effect 😅

Reviewer 2: The initial idea was to avoid having people linking to pages that are not too related to the steps or repeat the same link over and over. At that time it made sense to me but now I'm not too sure whether we should try to limit links. I actually found myself adding links like crazy in my tutorials and how-tos.
